Data/Atomic.hs

```diff
@@ -12,33 +13,52 @@ module Data.Atomic
     , subtract
     ) where
 
-import Data.Int (Int64)
-import Foreign.ForeignPtr (ForeignPtr, mallocForeignPtr, withForeignPtr)
-import Foreign.Ptr (Ptr)
-import Foreign.Storable (poke)
 import Prelude hiding (read, subtract)
 
+#include "MachDeps.h"
+
+-- We want an atomic with at least 64 bits in order to avoid overflow.
+
+#if WORD_SIZE_IN_BITS >= 64
```

Comment:
I'm curious about the need for optimizing this so carefully here. It's worth adding a comment at least about why we need to optimize this so much. My intuition is that for our needs, this could be premature optimization, given that the time to compute some estimators is probably a drop in the ocean, given the amount of JSON serialization and network IO that we're doing in the course of reporting this stuff. I could believe it's needed though, so any elucidation in comments here would be appreciated.

Reply:
Hmmmm. I don't have much knowledge or experience about optimization, so I might not be very helpful here. @jberryman, could I have your take on this?

The use of fetch-and-add here is inherited from the replaced `ekg-core` code, the implementation of which was motivated by reports of contention in `ekg` under the an old implementation that used `atomicModifyIORef'` -- see the last section of https://blog.johantibell.com/2014/05/ekg-04-released.html. Indeed, repeating Tibbel's contention-heavy micro-benchmark, `atomicModifyIORef` was over 100x slower than fetch-and-add, and about 5x as slow as [`atomicModifyIORefCas_`](https://hackage.haskell.org/package/atomic-primops-0.8.4/docs/Data-Atomics.html#v:atomicModifyIORefCAS_). It seems highly unlikely that our use case would trigger that level of contention, but there seems also seems to be little downside to using fetch-and-add here.

Padding the counter to 64 bytes is new -- I copied that trick from `unagi-chan`. However, unlike fetch-and-add, it has the downside that it uses additional memory, so this might be premature optimization. @jberryman was also skeptical about this padding when I showed to him a while back. In any case, it would do our use case no harm if we removed this padding.

Reply:
I've added commits removing the padding for the counter metric and justifing it for the distribution metric.